src/main/java/com/compilercharisma/chameleonbusinessstudio/controller/AppointmentController.java

```diff
@@ -1,163 +1,282 @@
+            @RequestParam("email") String email
+    ){
+        if(email == null){
+            return Mono.error(new IllegalArgumentException("Email cannot be null"));
+        }
+        if(!isValidEmail(email)){
+            return Mono.error(new IllegalArgumentException("Invalid email: " + email));
+        }
+
+        // Matt hates Reactor. Java needs Async/await!
+        return users.isRegistered(email)
+            .handle((Boolean isRegistered, SynchronousSink<Boolean> sink)->{
+                if(isRegistered){
+                    sink.next(isRegistered);
+                } else {
+                    sink.error(new IllegalArgumentException("Email is not registered as a user: " + email));
+                }
+            }).handle((Boolean whatever, SynchronousSink<AppointmentEntity> sink) -> {
+                var appointmentMaybe = appointments.getAppointmentById(appointmentId);
+                if(appointmentMaybe.isEmpty()){
+                    sink.error(new IllegalArgumentException("Invalid appointment ID: " + appointmentId));
+                } else {
+                    sink.next(appointmentMaybe.get());
+                }
+            }).handle((AppointmentEntity appt, SynchronousSink<AppointmentEntity> sink)->{
+                // need to call reactive from inside another reactive, pass down appt
+                authentication
+                    .getLoggedInUserReactive()
+                    .handle((user, userSink)->{
+                        // check if logged in user is allowed to book
+                        var role = user.getRole().getRoleName();
+                        if(isRoleAllowedToBookOtherPeople(role)){
+                            sink.next(appt);
+                        } else {
+                            var msg = String.format(
+                                "Users with role \"%s\" are not allowed to book appointments for other users",
+                                role
+                            );
+                            sink.error(new AccessDeniedException(msg));
+                        }
+                    });
+            }).handle((AppointmentEntity appt, SynchronousSink<AppointmentEntity> sink) -> {
+                if(appointments.isSlotAvailable(appt)){
+                    sink.next(appt);
+                } else {
+                    sink.error(new UnsupportedOperationException("Appointment is full; it cannot have any more participants"));
+                }
+            }).map(appt -> {
+                if(!appointments.isUserRegistered(appt, email)){
+                    appointments.registerUser(appt, email);
+                }
+        
+                return ResponseEntity.ok(appt);
+            });
+    }
+    
+    private boolean isRoleAllowedToCreateAppointments(String role){
+        var hs = new HashSet<>();
+        hs.add(Role.ADMIN);
+        hs.add(Role.ORGANIZER);
+        hs.add(Role.TALENT);
+        return hs.contains(role);
+    }
```

Comment:
Just a minor thing, we can also delegate this to an AuthorizationManager but if this works, then this will do.